### PR TITLE
Refs #17198 -- Detected existing total ordering in admin changelist.

### PR DESCRIPTION
**Pull request by charettes**
Appending pk is not necessary when the currently ordered by columns are part of a non-nullable unique contraint.

----

I didn't feel creating a new ticket was required given it was heavily discussed in [#17198](https://code.djangoproject.com/ticket/17198#comment:14) but I'd be happy to fill one if deemed necessary.

I decided to tackle this because one of the project I was working on was getting a really slow `/admin/auth/user` page due to the `username, -pk` ordering and `username` was performing way faster.

### CODE REVIEW

File: django/contrib/admin/views/main.py [hunk 1]
```diff
@@ -303,14 +303,57 @@ def get_ordering(self, request, queryset):
         # Add the given query's ordering fields, if any.
         ordering.extend(queryset.query.order_by)
 
-        # Ensure that the primary key is systematically present in the list of
+        # Ensure that a unique together set of fields is present in the list of
```
[the remark this answers is not in the document]
Comment (charettes) [reply django/contrib/admin/views/main.py]: I guess we could add a `get_total_ordering` method that gets passed `ordering` but I'm not sure if it shouldn't be the other way around: `get_queryset` calls `get_total_ordering` that calls `get_ordering`. It seems like it would make more sense than the `get_queryset(get_ordering(get_total_ordering))` chain.

Would you like me to create a new ticket for this feature and continue work in this PR?

[the remark this answers is not in the document]
Comment (carltongibson) [reply django/contrib/admin/views/main.py]: Errr... 😃 it’s part of the old ticket in my view. (That may not be canonical.)

A new ticket might help in case we need multiple reviews, but having come across this today, I kind of thought it would be good to get in for 2.2 if that’s realistic on you end. 

Tl;dr: don’t really mind. 

My initial thought on the code structure would be to go for the minimal change here. Add the helper (even private) to encapsulate this logic, and call it quits. We can always allow a restructuring later. 

However, if you have ideas for something better I wouldn’t object. 🙂

[the remark this answers is not in the document]
Comment (carltongibson) [reply django/contrib/admin/views/main.py]: OK, I see. 

I might be tempted to do it the other way round. Main reason is that if I override `get_total_ordering()` I need to call `get_ordering()` as the first step, whereas if `get_total_ordering()` just takes and returns the `ordering` it's freestanding. 

Not strongly inclined though. I'm happy to go with your preferred option. 

Do you want to document it, maybe as a `Fixed #29943 -- ...`?

**Comment by charettes** [comment]
@carltongibson I pushed an adjusted version of the patch that uses your favored approach but I didn't add documentation because I wasn't sure if this was going to make the 2.2 cut at this point. Let me know how you'd like me to proceed next.

**Comment by carltongibson** [comment]
Hey @charettes. Thanks. Lets see if we can sneak this in... 😬

I think I was getting too excited the other day: adding documentation for #29943 is better separate. 

Let me review the ordering docs this morning and we'll see if any docs changes are needed.

File: django/contrib/admin/views/main.py [hunk 2]
```diff
@@ -303,14 +303,66 @@ def get_ordering(self, request, queryset):
         # Add the given query's ordering fields, if any.
         ordering.extend(queryset.query.order_by)
 
-        # Ensure that the primary key is systematically present in the list of
-        # ordering fields so we can guarantee a deterministic order across all
+        return self.get_deterministic_ordering(ordering)
+
+    def get_deterministic_ordering(self, ordering):
+        """
+        Augmented the provided list of ordering with a descendant primary key
```
Comment (carltongibson) [inline django/contrib/admin/views/main.py]: Augment...

File: django/contrib/admin/views/main.py [hunk 3]
```diff
@@ -303,14 +303,66 @@ def get_ordering(self, request, queryset):
...
-            # The two sets do not intersect, meaning the pk isn't present. So
-            # we add it.
-            ordering.append('-pk')
+        return self._get_deterministic_ordering(ordering)
+
+    def _get_deterministic_ordering(self, ordering):
+        """
+        Ensure a deterministic order across all database backends. Search for a
+        single field or unique together set of fields providing a total
+        ordering. If these are missing augment the ordering with a descendant
```
Comment (timgraham) [inline django/contrib/admin/views/main.py]: "missing, "

File: django/contrib/admin/views/main.py [hunk 4]
```diff
@@ -303,14 +303,66 @@ def get_ordering(self, request, queryset):
...
+        primary key.
+        """
+        ordering = list(ordering)
+        ordering_fields = set()
+        total_ordering_fields = {'pk'}
+        total_ordering_fields.update(
+            field.attname for field in self.lookup_opts.fields if field.unique and not field.null
```
Comment (timgraham) [inline django/contrib/admin/views/main.py]: I would multiline:
```
field.attname for field in self.lookup_opts.fields
if field.unique and not field.null
```

File: django/contrib/admin/views/main.py [hunk 5]
```diff
@@ -303,14 +303,66 @@ def get_ordering(self, request, queryset):
...
+                    continue
+                if field.attname in total_ordering_fields:
+                    break
+                ordering_fields.add(field.attname)
+        else:
+            # No single total ordering field, try unique_together.
+            for field_names in self.lookup_opts.unique_together:
+                # Normalize attname references by using get_field().
+                fields = [
+                    self.lookup_opts.get_field(field_name) for field_name in field_names
```
Comment (timgraham) [inline django/contrib/admin/views/main.py]: I would single line: `fields = [...]` or multiline as in previous comment.

File: django/contrib/admin/views/main.py [hunk 6]
```diff
@@ -303,14 +303,66 @@ def get_ordering(self, request, queryset):
...
+                ]
+                # Composite unique constraints containing a nullable column
+                # cannot ensure total ordering.
+                if any(field.null for field in fields):
+                    continue
+                field_attnames = {field.attname for field in fields}
+                if ordering_fields.issuperset(field_attnames):
+                    break
+            else:
+                # If no set of unique fields is present in the ordering we rely on
```
Comment (timgraham) [inline django/contrib/admin/views/main.py]: "ordering, rely" (add comma & chop "we")

File: tests/admin_changelist/tests.py [hunk 7]
```diff
@@ -937,6 +939,89 @@ def check_results_order(ascending=False):
...
+            null_field = models.BooleanField(null=True)
+
+            class Meta:
+                unique_together = {
+                    ('field', 'other_field'),
+                    ('field', 'null_field'),
+                    ('related', 'other_related_id')
+                }
+
+        queryset = Model.objects.none()
```
Comment (timgraham) [inline tests/admin_changelist/tests.py]: I there a reason for the variable as opposed to `return Model.objects.none()`?

File: tests/admin_changelist/tests.py [hunk 8]
```diff
@@ -937,6 +939,89 @@ def check_results_order(ascending=False):
...
+            related_unique = models.OneToOneField(Related, models.CASCADE)
+            field = models.BooleanField()
+            other_field = models.BooleanField()
+            null_field = models.BooleanField(null=True)
+
+            class Meta:
+                unique_together = {
+                    ('field', 'other_field'),
+                    ('field', 'null_field'),
+                    ('related', 'other_related_id')
```
Comment (timgraham) [inline tests/admin_changelist/tests.py]: add trailing comma

File: tests/admin_changelist/tests.py [hunk 9]
```diff
@@ -937,6 +939,89 @@ def check_results_order(ascending=False):
...
+
+        class ModelAdmin(admin.ModelAdmin):
+            def get_queryset(self, request):
+                return queryset
+
+        request = self._mocked_authenticated_request('/', self.superuser)
+        site = admin.AdminSite(name='admin')
+        model_admin = ModelAdmin(Model, site)
+        change_list = model_admin.get_changelist_instance(request)
+        self.assertEqual(change_list._get_deterministic_ordering([]), ['-pk'])
```
Comment (timgraham) [inline tests/admin_changelist/tests.py]: Did you consider a loop with subTest for all these assertions?

Comment (carltongibson) [reply tests/admin_changelist/tests.py]: TBH, didn't think about it. 

I quite liked the presentation (with the comments) when first reading it through: helped me reason about it.

**Comment by carltongibson** [comment]
I addressed all but one of Tim's points. (I think I broke the Windows build pushing too quickly...) 

Remaining point is re subTest() — will comment inline. 

Bar this, I'm ready to go rolling v2.2a1.

File: tests/admin_changelist/tests.py (hunk 9 above)
Comment (timgraham) [reply tests/admin_changelist/tests.py]: Sure, we sometimes use comments with that pattern;
```
tests = (
    # Unique non-nullable field.
    (['unique_field']), ['unique_field']),
    ...
    # ....
    ...
)
```

Comment (carltongibson) [reply tests/admin_changelist/tests.py]: OK. That makes sense. Let me have a go at re-writing that.

File: django/contrib/admin/views/main.py (hunk 4 above)
Comment (pope1ni) [reply django/contrib/admin/views/main.py]:
```python
total_ordering_fields = {'pk', *(
    field.attname for field in self.lookup_opts.fields
    if field.unique and not field.null
)}
```
?

File: django/contrib/admin/views/main.py [hunk 10]
```diff
@@ -303,14 +303,65 @@ def get_ordering(self, request, queryset):
...
+        ordering. If these are missing, augment the ordering with a descendant
+        primary key.
+        """
+        ordering = list(ordering)
+        ordering_fields = set()
+        total_ordering_fields = {'pk'}
+        total_ordering_fields.update(
+            field.attname for field in self.lookup_opts.fields
+            if field.unique and not field.null
+        )
```
Comment (pope1ni) [inline django/contrib/admin/views/main.py]:
```python
total_ordering_fields = {'pk'} | {
    field.attname for field in self.lookup_opts.fields
    if field.unique and not field.null
}
```

File: django/contrib/admin/views/main.py [hunk 11]
```diff
@@ -303,14 +303,65 @@ def get_ordering(self, request, queryset):
...
+        for part in ordering:
+            # Search for single field providing a total ordering.
+            field_name = None
+            if isinstance(part, str):
+                if part.startswith('-'):
+                    field_name = part[1:]
+                else:
+                    field_name = part
```
Comment (pope1ni) [inline django/contrib/admin/views/main.py]: May as well do the following as a field name can only legally have a single `-` at the start:
```python
field_name = part.lstrip('-')
```

File: django/contrib/admin/views/main.py [hunk 12]
```diff
@@ -303,14 +303,65 @@ def get_ordering(self, request, queryset):
...
+            # No single total ordering field, try unique_together.
+            for field_names in self.lookup_opts.unique_together:
+                # Normalize attname references by using get_field().
+                fields = [self.lookup_opts.get_field(field_name) for field_name in field_names]
+                # Composite unique constraints containing a nullable column
+                # cannot ensure total ordering.
+                if any(field.null for field in fields):
+                    continue
+                field_attnames = {field.attname for field in fields}
+                if ordering_fields.issuperset(field_attnames):
```
Comment (pope1ni) [inline django/contrib/admin/views/main.py]:
```python
if ordering_fields.issuperset(field.attname for field in fields):
```

File: tests/admin_changelist/tests.py (hunk 9 above)
Comment (carltongibson) [reply tests/admin_changelist/tests.py]: https://github.com/django/django/pull/10692/commits/deadb225ccb703b24ee685e079930541149e2370 (which is a `fixup`)

**Comment by carltongibson** [comment]
Super. Thanks @pope1ni. They're in https://github.com/django/django/pull/10692/commits/6d34b01186061c1ef826c036c843c8b0f1a5889b

**Comment by carltongibson** [comment]
Again, with breaking the Windows build. I'll give it 10 minutes then squash...

File: tests/admin_changelist/tests.py [hunk 13]
```diff
@@ -937,6 +939,85 @@ def check_results_order(ascending=False):
...
+
+        class ModelAdmin(admin.ModelAdmin):
+            def get_queryset(self, request):
+                return Model.objects.none()
+
+        request = self._mocked_authenticated_request('/', self.superuser)
+        site = admin.AdminSite(name='admin')
+        model_admin = ModelAdmin(Model, site)
+        change_list = model_admin.get_changelist_instance(request)
+
```
Comment (timgraham) [inline tests/admin_changelist/tests.py]: All the blank lines don't add much for me but I won't insist.

File: tests/admin_changelist/tests.py [hunk 14]
```diff
@@ -937,6 +939,85 @@ def check_results_order(ascending=False):
...
+        # F() objects composite unique nullable.
+        non_total_ordering = [F('field'), F('null_field').desc(nulls_last=True)]
+
+        tests += (
+            (total_ordering, total_ordering),
+            (non_total_ordering, non_total_ordering + ['-pk']),
+        )
+
+        for ordering, expected in tests:
+            with self.subTest(expected=expected):
```
Comment (timgraham) [inline tests/admin_changelist/tests.py]: I think this should use `ordering` because `expected` will already appear in the output if the assertion fails.

File: django/contrib/admin/views/main.py [hunk 15]
```diff
@@ -303,14 +303,60 @@ def get_ordering(self, request, queryset):
...
+                # Composite unique constraints containing a nullable column
+                # cannot ensure total ordering.
+                if any(field.null for field in fields):
+                    continue
+                if ordering_fields.issuperset(field.attname for field in fields):
+                    break
+            else:
+                # If no set of unique fields is present in the ordering, rely
+                # on the primary key to provide total ordering.
+                ordering.append('-pk')
```
Comment (timgraham) [inline django/contrib/admin/views/main.py]: I'd remove this blank line.

File: tests/admin_changelist/tests.py (hunk 13 above)
Comment (carltongibson) [reply tests/admin_changelist/tests.py]: Fine. They're gone. 😀

File: tests/admin_changelist/tests.py (hunk 14 above)
Comment (carltongibson) [reply tests/admin_changelist/tests.py]: OK, good. Done.

**Comment by charettes** [comment]
Thanks for the adjustments and merge @carltongibson, @pope1ni, and @timgraham, really appreciate you took the time to sneak this one in.
